lnwallet/chanfunding/coin_select_test.go

```diff
@@ -505,28 +531,215 @@ func TestCoinSelectSubtractFees(t *testing.T) {
+	t.Parallel()
+
+	const feeRate = chainfee.SatPerKWeight(100)
+	const dustLimit = btcutil.Amount(1000)
+	const dust = btcutil.Amount(100)
+	const coin = btcutil.SatoshiPerBitcoin
+
+	type testCase struct {
+		name     string
+		maxValue btcutil.Amount
+		coins    []Coin
+
+		expectedInput      []btcutil.Amount
+		expectedFundingAmt btcutil.Amount
+		expectedChange     btcutil.Amount
+		expectErr          bool
+	}
+
+	testCases := []testCase{{
+		// We have 1.0 BTC available, spend them all.
+		// This should lead to a funding TX with one output, the rest
+		// goes to fees.
+		name: "spend exactly all",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 1 * coin,
+
+		// The one and only input will be selected.
+		expectedInput:      []btcutil.Amount{1 * coin},
+		expectedFundingAmt: 1*coin - fundingFee(feeRate, 1, false),
+		expectedChange:     0,
+	}, {
+		// We have 1.0 BTC available and want to spend up to 2 BTC.
+		// This should lead to a funding TX with one output, the rest
+		// goes to fees.
+		name: "spend more",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 2 * coin,
+
+		// The one and only input will be selected.
+		expectedInput:      []btcutil.Amount{1 * coin},
+		expectedFundingAmt: 1*coin - fundingFee(feeRate, 1, false),
+		expectedChange:     0,
+	}, {
+		// We have 1.0 BTC available and want to spend up to 0.5 BTC.
+		// This should lead to a funding TX with one output and a
+		// change to subtract the fees from.
+		name: "spend far below",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 0.5 * coin,
+
+		// The one and only input will be selected.
+		expectedInput:      []btcutil.Amount{1 * coin},
+		expectedFundingAmt: 0.5 * coin,
+		expectedChange:     0.5*coin - fundingFee(feeRate, 1, true),
+	}, {
+		// We have 1.0 BTC available and want to spend just 1 Satoshi
+		// below that amount.
+		// This should lead to a funding TX with one output where the
+		// fee is subtracted from the total 1 BTC input value.
+		name: "spend little below",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 1*coin - 1,
+
+		// The one and only input will be selected.
+		expectedInput: []btcutil.Amount{
+			1 * coin,
+		},
+		expectedFundingAmt: 1*coin - fundingFee(feeRate, 1, false),
+		expectedChange:     0,
+	}, {
+		// The total funds available is below the dust limit after
+		// paying fees.
+		name: "dust output",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value: int64(
+					fundingFee(feeRate, 1, false) + dust,
+				),
+			},
+		}},
+		maxValue: fundingFee(feeRate, 1, false) + dust,
+
+		expectErr: true,
+	}, {
+		// If more than 20% of funds goes to fees, it should fail.
+		name: "high fee",
```

Comment:
this test fails not because of high-fees, but because of the dust-limit so still the test missing where we abort because of high fees

Reply:
I think we should check the error type in the tests to make sure we test we want to test to prevent such cases ?

Reply:
Very good catch! I've adjusted this to run into the high fee case and also added the comparison of error strings.